[FLINK-35167][cdc-connector] Introduce MaxCompute pipeline DataSink

dingxin-tech opened the pull request:
By integrating the MaxCompute DataSink, we enable the precise and efficient synchronization of data from Flink's Change Data Capture (CDC) into MaxCompute.

kevinwangcs commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/sink/MaxComputeEventWriter.java:
```diff
@@ -0,0 +1,175 @@
+import java.util.concurrent.ExecutionException;
+
+/** a {@link SinkWriter} for {@link Event} for MaxCompute. */
+public class MaxComputeEventWriter implements SinkWriter<Event> {
```
Could you please consider changing "Event" to a generic type to support various serializer implementations, similar to the StarRocksSink [EventRecordSerializationSchema](https://github.com/apache/flink-cdc/blob/master/flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-starrocks/src/main/java/org/apache/flink/cdc/connectors/starrocks/sink/EventRecordSerializationSchema.java) and [DorisEventSerializer](https://github.com/apache/flink-cdc/blob/master/flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-doris/src/main/java/org/apache/flink/cdc/connectors/doris/sink/DorisEventSerializer.java) implementations in Flink CDC?

dingxin-tech replied:
I am very willing to provide a sink that supports generic types. However, the current code framework would require significant changes to implement this feature. This functionality may appear in the future in flink-connector-maxcompute rather than here in flink-cdc-connector.

loserwang1024 commented:
I am wondering how a commercial database sink like MaxCompute to do e2e test?

dingxin-tech commented:
> I am wondering how a commercial database sink like MaxCompute to do e2e test?

I will soon be working on creating a Docker image for a `MaxCompute Emulator` that launches a mocked version of MaxCompute. This will allow for end-to-end testing to be performed by initializing this image prior to regression testing.

dingxin-tech commented:
hi @loserwang1024 , I recently completed the development and release of a Docker image for the MaxCompute Emulator, and I have also added some related regression cases. Could you help me review the code in this pull request?

lvyanquan reviewed (commented):
Thanks for your contribution, I've left some comments.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/sink/MaxComputeEventWriter.java:
```diff
@@ -0,0 +1,197 @@
+
+    @Override
+    public void write(Event element, Context context) throws IOException {
+        LOG.info("Sink writer {} write {}.", this.context.getSubtaskId(), element);
```
Sure, I forget to remove this log for debug use.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/sink/MaxComputeEventSink.java:
```diff
@@ -0,0 +1,82 @@
+        //                                "PostPartition",
+        //                                new EventTypeInfo(),
+        //                                new PostPartitionOperator(stream.getParallelism()))
+        //                        .name("PartitionByBucket");
```
Sorry, I comment out this part of code for debug use but forget to un-comment.
I think this code will stay here until the runtime optimization your mentioned is passed, and then I will re-implement this feature in that way.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/test/java/org/apache/flink/cdc/connectors/maxcompute/utils/SchemaEvolutionUtilsTest.java:
```diff
@@ -0,0 +1,97 @@
+import org.junit.jupiter.api.Test;
+
+/** e2e test of SchemaEvolutionUtils. */
+public class SchemaEvolutionUtilsTest {
```
Sure.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/utils/TypeConvertUtils.java:
```diff
@@ -0,0 +1,540 @@
+            case CHAR:
+            case VARCHAR:
+            case TIME_WITHOUT_TIME_ZONE:
+                return TypeInfoFactory.STRING;
```
Yes, you are correct.

Additionally, I discovered that the MaxCompute SDK has an issue with creating tables based on primary keys. This issue results in ignoring the user-specified primary key order during table creation. I plan to fix this next week.

dingxin-tech commented on docs/content.zh/docs/connectors/maxcompute.md:
```diff
@@ -0,0 +1,342 @@
+      <td>optional</td>
+      <td style="word-wrap: break-word;">16</td>
+      <td>Integer</td>
+      <td>自动创建 MaxCompute Transaction 表时使用的桶数。使用方式可以参考 <a href="ttps://help.aliyun.com/zh/maxcompute/user-guide/table-data-format">
```
Fixed.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/resources/META-INF/services/org.apache.flink.cdc.common.factories.Factory:
```diff
@@ -0,0 +1,17 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+org.apache.flink.cdc.connectors.maxcompute.MaxComputeDataSinkFactory
```
Done.

lvyanquan commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/coordinator/SessionManageOperator.java:
```diff
@@ -0,0 +1,297 @@
+    @Override
+    public void snapshotState(StateSnapshotContext context) throws Exception {
+        super.snapshotState(context);
+        sessionCache.clear();
```
So, do we need to request a new session ID after each checkpoint, which may have a performance impact? It is expected that the checkpoint interval will need to be larger, right.

dingxin-tech replied:
Yes, we do. 
Each session can no longer be used after being committed, so we must re-create a session and request a new sessionID. 
And indeed, the checkpoint interval will need to be larger.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/utils/SessionCommitCoordinator.java:
```diff
@@ -0,0 +1,158 @@
+ * completion status of an executor, allowing the system to determine whether all relevant sessions
+ * have been processed.
+ */
+public class SessionCommitCoordinator {
```
Sure, I didn't think of such a suitable name when I was naming it.

leonardBang reviewed (commented):
Thanks @dingxin-tech for the nice work, the CI wasn't triggered properly, we need adjust the CI setting[1] when new connector or new module joining.

[1]https://github.com/apache/flink-cdc/blob/master/.github/workflows/flink_cdc.yml

dingxin-tech commented:
> Thanks @dingxin-tech for the nice work, the CI wasn't triggered properly, we need adjust the CI setting[1] when new connector or new module joining.
> 
> [1]https://github.com/apache/flink-cdc/blob/master/.github/workflows/flink_cdc.yml

Hi, I added tests for MaxCompute in the CI file. Additionally, I refactored the code to apply the newly released DataSink feature, which allows specifying a HashFunction, and upgraded the ODPS SDK. Could you please review this pr again?
@leonardBang @lvyanquan

lvyanquan commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/coordinator/SessionManageOperator.java:
```diff
@@ -0,0 +1,297 @@
+    private static final Logger LOG = LoggerFactory.getLogger(SessionManageOperator.class);
+
+    /** a tricky way to get an Operator from sink. */
+    public static SessionManageOperator instance;
```
We should add `todo` to note that we can try to remove it in the future.

dingxin-tech replied:
ok.

lvyanquan commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/utils/MaxComputeUtils.java:
```diff
@@ -0,0 +1,258 @@
+
+    public static CompressOption compressOptionOf(String compressAlgo) {
+        CompressOption.CompressAlgorithm compressAlgorithm;
+        switch (compressAlgo) {
```
What about using CompressOption.CompressAlgorithm.valueOf() to simplify this?

dingxin-tech replied:
Perhaps using zlib and lz4 instead of odps_zlib and odps_lz4_frame would be clearer for users?

lvyanquan reviewed (commented):
Thanks @dingxin-tech  for these updates, I left some comments.

lvyanquan commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/sink/MaxComputeEventWriter.java:
```diff
@@ -0,0 +1,190 @@
+
+    @Override
+    public void flush(boolean endOfInput) throws IOException, InterruptedException {
+        SessionManageOperator operator = SessionManageOperator.instance;
```
We can add Preconditions.checkNotNull() check here to guide users avoid setting configuration like 'pipeline.operator-chaining'='false'

dingxin-tech replied:
ok

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/MaxComputeMetadataApplier.java:
```diff
@@ -0,0 +1,112 @@
+                        addColumnEvent.tableId(),
+                        addColumnEvent.getAddedColumns());
+            } else {
+                throw new UnsupportedOperationException(
```
Yes, I will support it.

dingxin-tech commented on flink-cdc-connect/flink-cdc-pipeline-connectors/flink-cdc-pipeline-connector-maxcompute/src/main/java/org/apache/flink/cdc/connectors/maxcompute/utils/SchemaEvolutionUtils.java:
```diff
@@ -0,0 +1,256 @@
+     */
+    public static void alterColumnType(
+            MaxComputeOptions options, TableId tableId, Map<String, DataType> typeMapping)
+            throws OdpsException {
```
Done.

lvyanquan commented on flink-cdc-e2e-tests/flink-cdc-pipeline-e2e-tests/src/test/java/org/apache/flink/cdc/pipeline/tests/MaxComputeE2eITCase.java:
```diff
@@ -0,0 +1,193 @@
+                        "select * from table1 order by col1;");
+        instance.waitForSuccess();
+        List<Record> result = SQLTask.getResult(instance);
+        System.out.println(result);
```
Can be removed.

lvyanquan commented on flink-cdc-e2e-tests/flink-cdc-pipeline-e2e-tests/src/test/java/org/apache/flink/cdc/pipeline/tests/MaxComputeE2eITCase.java:
```diff
@@ -0,0 +1,193 @@
+        Assert.assertEquals("x", result.get(0).get(1));
+        // 3, NULL (MaxCompute Emulator use 'NULL' instead of null)
+        Assert.assertEquals("3", result.get(1).get(0));
+        Assert.assertEquals("NULL", result.get(1).get(1));
```
What about add more verification of insert/update records and schema change after snapshot reading.

lvyanquan commented on flink-cdc-e2e-tests/flink-cdc-pipeline-e2e-tests/src/test/java/org/apache/flink/cdc/pipeline/tests/MaxComputeE2eITCase.java:
```diff
@@ -0,0 +1,193 @@
+
+        String pipelineJob =
+                "source:\n"
+                        + "   type: values\n"
```
It's better to use MySQL as source connector like MySqlToDorisE2eITCase and try to cover more types.

lvyanquan reviewed (approved):
LGTM, left some comments about MaxComputeE2eITCase.

leonardBang commented:
@dingxin-tech Could you rebase your PR to latest master branch? btw the PR description is required

dingxin-tech commented:
> @dingxin-tech Could you rebase your PR to latest master branch? btw the PR description is required

Done.

leonardBang reviewed (approved):
Thanks @dingxin-tech for the update and @lvyanquan  for the review work, CI passed, merging...
